File: server/youtubePlaylist.js
/**
 * Server-side YouTube playlist resolver for The Orpheum's import flow.
 *
 * Browsers cannot fetch youtube.com (CORS), so the server reads a public
 * playlist page and extracts its video list from the embedded ytInitialData
 * JSON. No API key, no new dependency: one bounded fetch plus a pure
 * extraction core that the node tests pin against synthetic fixtures.
 *
 * Rules (mirroring the IPTV fetch): http(s) by construction, one 15 s
 * timeout, a streaming size cap, and a hard cap on resolved videos. Mixes
 * and radios (infinite lists) are declined, not truncated silently.
 * Extraction robustness is a known risk (YouTube markup drift); every
 * failure maps to a stable reason string the shared model translates.
 */

import { THEATER_LIMITS } from '../shared/theaterModel.js';

const FETCH_TIMEOUT_MS = 15_000;
const FETCH_MAX_BYTES = 3 * 1024 * 1024; // a playlist page is big; cap tightly anyway
const PLAYLIST_ID_RE = /^[\w-]{12,}$/;

/**
 * Mix/radio ids never end: RD… (radio) and UL… (My Mix) are generated
 * queues, not fixed playlists, so there is no bounded list to import.
 */
export function isYouTubeMixId(listId) {
  return /^(RD|UL)/.test(String(listId || ''));
}

/** True when the string looks like a real playlist id (PL…, OLAK5uy_…, UU…). */
export function looksLikePlaylistId(listId) {
  return typeof listId === 'string' && PLAYLIST_ID_RE.test(listId);
}

/**
 * Pull the first ytInitialData assignment out of the page HTML and parse
 * it. Returns the data object or null (consent walls, unreadable pages).
 */
function parseYtInitialData(html) {
  const marker = html.indexOf('ytInitialData');
  if (marker === -1) return null;
  const start = html.indexOf('{', marker);
  if (start === -1) return null;
  // Walk the JSON text itself: brace matching respects strings and escapes,
  // so no regex gamble against a megabyte of markup.
  let depth = 0;
  let inString = false;
  let escaped = false;
  for (let i = start; i < html.length; i++) {
    const ch = html[i];
    if (inString) {
      if (escaped) escaped = false;
      else if (ch === '\\') escaped = true;
      else if (ch === '"') inString = false;
      continue;
    }
    if (ch === '"') inString = true;
    else if (ch === '{') depth++;
    else if (ch === '}') {
      depth--;
      if (depth === 0) {
        try {
          return JSON.parse(html.slice(start, i + 1));
        } catch {
          return null;
        }
      }
    }
  }
  return null;
}

/** Playlist title from the metadata blocks, or null. */
function extractPlaylistTitle(data) {
  const title =
    data?.metadata?.playlistMetadataRenderer?.title ||
    data?.microformat?.microformatDataRenderer?.title ||
    data?.header?.playlistHeaderRenderer?.title;
  const text = typeof title === 'string'
    ? title
    : Array.isArray(title?.runs) ? title.runs.map((r) => r?.text).join('') : title?.simpleText;
  return typeof text === 'string' ? text.trim() : null;
}

/** {videoId, title} from a playlist-entry node, or null. YouTube ships two
 * layouts for the same thing: the classic playlistVideoRenderer and the
 * newer lockupViewModel (contentId + metadata title content). */
function videoFromNode(node) {
  if (node?.playlistVideoRenderer) {
    const r = node.playlistVideoRenderer;
    const videoId = typeof r?.videoId === 'string' ? r.videoId : null;
    if (!videoId || !/^[\w-]{6,}$/.test(videoId)) return null;
    const title = Array.isArray(r?.title?.runs)
      ? r.title.runs.map((t) => t?.text).join('')
      : r?.title?.simpleText;
    return { videoId, title: typeof title === 'string' ? title.trim() : '' };
  }
  if (node?.lockupViewModel) {
    const l = node.lockupViewModel;
    // Only plain video lockups are playlist entries; playlist lockups carry
    // a list id in contentId and would smuggle an unplayable entry in.
    if (typeof l?.contentType === 'string' && l.contentType !== 'LOCKUP_CONTENT_TYPE_VIDEO') return null;
    const videoId = typeof l?.contentId === 'string' ? l.contentId : null;
    if (!videoId || !/^[\w-]{6,}$/.test(videoId)) return null;
    const title = l?.metadata?.lockupMetadataViewModel?.title?.content;
    return { videoId, title: typeof title === 'string' ? title.trim() : '' };
  }
  return null;
}

/**
 * Depth-first sweep for playlist-entry nodes (playlistVideoRenderer on the
 * classic layout, lockupViewModel on the current one), in page order.
 * Sweeping beats hard-coding one container path.
 */
function collectVideos(data) {
  const videos = [];
  const seen = new Set();
  const visit = (node) => {
    if (!node || typeof node !== 'object' || videos.length >= THEATER_LIMITS.RESOLVE_MAX) return;
    if (Array.isArray(node)) {
      for (const child of node) visit(child);
      return;
    }
    const video = videoFromNode(node);
    if (video) {
      if (!seen.has(video.videoId)) {
        seen.add(video.videoId);
        videos.push(video);
      }
      return; // entry nodes are leaves for our purposes
    }
    for (const child of Object.values(node)) {
      if (child && typeof child === 'object') visit(child);
    }
  };
  visit(data);
  return videos;
}

/**
 * Pure extraction core: playlist HTML -> `{ title, videos:
 * [{videoId, title}] }` (ordered, deduped, capped at RESOLVE_MAX) or
 * `{ reason }` for a page that is not a readable public playlist. Synthetic
 * fixtures in the tests pin the shape.
 */
export function extractPlaylistVideos(html) {
  if (typeof html !== 'string' || !html.includes('ytInitialData')) {
    return { reason: 'playlist_unreadable' };
  }
  const data = parseYtInitialData(html);
  if (!data) return { reason: 'playlist_unreadable' };
  const videos = collectVideos(data);
  if (!videos.length) {
    // A real page that names no videos: private/deleted playlists look
    // exactly like this (alertsRenderer instead of a video list).
    return { reason: 'playlist_not_public' };
  }
  return {
    title: extractPlaylistTitle(data) || 'A YouTube playlist',
    videos,
  };
}

/**
 * Bounded resolve of one playlist id: fetch the page and extract. Returns
 * `{ title, videos }` or `{ reason }` — the same shapes as the pure core,
 * plus network-level failures mapped to stable reasons.
 */
export async function resolvePlaylist(listId) {
  if (!looksLikePlaylistId(listId)) return { reason: 'playlist_unreadable' };
  if (isYouTubeMixId(listId)) return { reason: 'is_mix' };

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), FETCH_TIMEOUT_MS);
  try {
    // hl=en pins the extraction language; the consent cookies skip the EU
    // interstitial that would otherwise serve a page with no ytInitialData.
    const res = await fetch(`https://www.youtube.com/playlist?list=${encodeURIComponent(listId)}&hl=en`, {
      signal: controller.signal,
      redirect: 'follow',
      headers: {
        'accept-language': 'en',
        cookie: 'CONSENT=YES+cb; SOCS=CAI',
      },
    });
    if (res.status === 403 || res.status === 404) return { reason: 'playlist_not_public' };
    if (!res.ok) return { reason: 'playlist_unreadable' };
    const declared = Number(res.headers.get('content-length'));
    if (Number.isFinite(declared) && declared > FETCH_MAX_BYTES) return { reason: 'playlist_unreadable' };

    // Stream with a hard cap: the data we need sits well inside the first
    // megabytes, so an overgrown page is cut, not swallowed whole.
    const reader = res.body?.getReader?.();
    let html;
    if (reader) {
      const chunks = [];
      let size = 0;
      for (;;) {
        const { done, value } = await reader.read();
        if (done) break;
        size += value.length;
        if (size > FETCH_MAX_BYTES) {
          try {
            await reader.cancel();
          } catch {}
          return { reason: 'playlist_unreadable' };
        }
        chunks.push(value);
      }
      html = Buffer.concat(chunks).toString('utf8');
    } else {
      html = await res.text();
      if (html.length > FETCH_MAX_BYTES) return { reason: 'playlist_unreadable' };
    }
    return extractPlaylistVideos(html);
  } catch (err) {
    // Timeouts and network errors read the same to the player: try again.
    return { reason: 'playlist_unreadable' };
  } finally {
    clearTimeout(timer);
  }
}

File: server_elixir/lib/afterlight/theater/youtube_playlist.ex
defmodule Afterlight.Theater.YoutubePlaylist do
  @moduledoc """
  Server-side YouTube playlist resolver for The Orpheum import flow.

  Pure extraction core plus a bounded Finch fetch (15 s, 3 MiB streamed cap,
  `RESOLVE_MAX` 100). Mix ids `RD*` / `UL*` are declined at resolve time.
  """

  alias Afterlight.Theater.Reducer

  @fetch_timeout_ms 15_000
  @fetch_max_bytes 3 * 1024 * 1024
  @resolve_max Reducer.resolve_max()
  @playlist_id_re ~r/^[A-Za-z0-9_-]{12,}$/
  @video_id_re ~r/^[\w-]{6,}$/

  def fetch_timeout_ms, do: @fetch_timeout_ms
  def fetch_max_bytes, do: @fetch_max_bytes
  def resolve_max, do: @resolve_max

  def is_mix_id?(list_id) do
    list_id = to_string(list_id || "")
    String.match?(list_id, ~r/^(RD|UL)/)
  end

  def looks_like_playlist_id?(list_id) do
    is_binary(list_id) and Regex.match?(@playlist_id_re, list_id)
  end

  @doc """
  Bounded resolve of one playlist id. Returns `{:ok, %{title, videos}}` or
  `{:error, reason}` with stable reason strings.
  """
  def resolve(list_id) when is_binary(list_id) do
    cond do
      is_mix_id?(list_id) ->
        {:error, "is_mix"}

      not looks_like_playlist_id?(list_id) ->
        {:error, "playlist_unreadable"}

      true ->
        fetch_and_extract(list_id)
    end
  end

  def resolve(_), do: {:error, "playlist_unreadable"}

  @doc """
  Pure extraction: playlist HTML → `{:ok, %{title, videos}}` or `{:error, reason}`.
  """
  def extract_playlist_videos(html) when is_binary(html) do
    if String.contains?(html, "ytInitialData") do
      case parse_yt_initial_data(html) do
        :error ->
          {:error, "playlist_unreadable"}

        {:ok, data} ->
          case collect_videos(data) do
            [] -> {:error, "playlist_not_public"}
            videos -> {:ok, %{title: playlist_title(data) || "A YouTube playlist", videos: videos}}
          end
      end
    else
      {:error, "playlist_unreadable"}
    end
  end

  def extract_playlist_videos(_), do: {:error, "playlist_unreadable"}

  defp fetch_and_extract(list_id) do
    url = "https://www.youtube.com/playlist?list=#{URI.encode_www_form(list_id)}&hl=en"

    headers = [
      {"accept-language", "en"},
      {"cookie", "CONSENT=YES+cb; SOCS=CAI"}
    ]

    request = Finch.build(:get, url, headers)

    task =
      Task.async(fn ->
        Finch.request(request, Afterlight.Finch, receive_timeout: @fetch_timeout_ms)
      end)

    try do
      case Task.await(task, @fetch_timeout_ms + 1_000) do
        {:ok, %{status: status}} when status in [403, 404] ->
          {:error, "playlist_not_public"}

        {:ok, %{status: status, headers: resp_headers, body: body}} when status in 200..299 ->
          declared =
            resp_headers
            |> Enum.find_value(fn
              {"content-length", v} -> parse_int(v)
              {_, _} -> nil
            end)

          if is_integer(declared) and declared > @fetch_max_bytes do
            {:error, "playlist_unreadable"}
          else
            read_body(body)
          end

        {:ok, %{status: _}} ->
          {:error, "playlist_unreadable"}

        {:error, _} ->
          {:error, "playlist_unreadable"}
      end
    catch
      :exit, _ -> {:error, "playlist_unreadable"}
    end
  end

  defp read_body(body) when is_binary(body) do
    if byte_size(body) > @fetch_max_bytes do
      {:error, "playlist_unreadable"}
    else
      extract_playlist_videos(body)
    end
  end

  defp read_body({:stream, stream}) do
    stream
    |> Enum.reduce_while({:ok, []}, fn chunk, {:ok, acc} ->
      size = Enum.reduce(acc, byte_size(chunk), fn c, n -> n + byte_size(c) end)

      if size > @fetch_max_bytes do
        {:halt, {:error, "playlist_unreadable"}}
      else
        {:cont, {:ok, [chunk | acc]}}
      end
    end)
    |> case do
      {:ok, chunks} -> extract_playlist_videos(IO.iodata_to_binary(Enum.reverse(chunks)))
      {:error, reason} -> {:error, reason}
    end
  end

  defp read_body(_), do: {:error, "playlist_unreadable"}

  defp parse_int(v) when is_binary(v) do
    case Integer.parse(v) do
      {n, _} -> n
      :error -> nil
    end
  end

  defp parse_int(_), do: nil

  defp parse_yt_initial_data(html) do
    case :binary.match(html, "ytInitialData") do
      :nomatch ->
        :error

      {marker, _} ->
        <<_::binary-size(marker), after_marker::binary>> = html

        case :binary.match(after_marker, "{") do
          :nomatch ->
            :error

          {open, _} ->
            <<_::binary-size(open), json::binary>> = after_marker
            scan_json(json, 0, false, false, [])
        end
    end
  end

  defp scan_json(<<>>, _depth, _in_string, _escaped, _acc), do: :error

  defp scan_json(<<ch, rest::binary>>, depth, in_string, escaped, acc) do
    cond do
      in_string ->
        cond do
          escaped -> scan_json(rest, depth, true, false, [ch | acc])
          ch == ?\\ -> scan_json(rest, depth, true, true, [ch | acc])
          ch == ?" -> scan_json(rest, depth, false, false, [ch | acc])
          true -> scan_json(rest, depth, true, false, [ch | acc])
        end

      ch == ?" ->
        scan_json(rest, depth, true, false, [ch | acc])

      ch == ?{ ->
        scan_json(rest, depth + 1, false, false, [ch | acc])

      ch == ?} ->
        if depth == 1 do
          json = IO.iodata_to_binary(Enum.reverse(["}" | acc]))

          case Jason.decode(json) do
            {:ok, data} -> {:ok, data}
            _ -> :error
          end
        else
          scan_json(rest, depth - 1, false, false, [ch | acc])
        end

      true ->
        scan_json(rest, depth, false, false, [ch | acc])
    end
  end

  defp playlist_title(data) do
    title =
      get_in(data, ["metadata", "playlistMetadataRenderer", "title"]) ||
        get_in(data, ["microformat", "microformatDataRenderer", "title"]) ||
        get_in(data, ["header", "playlistHeaderRenderer", "title"])

    text =
      cond do
        is_binary(title) -> title
        is_list(get_in(title || %{}, ["runs"])) ->
          title["runs"]
          |> Enum.map(&to_string(Map.get(&1 || %{}, "text")))
          |> Enum.join("")

        is_map(title) -> title["simpleText"]
        true -> nil
      end

    if is_binary(text), do: String.trim(text), else: nil
  end

  defp video_from_node(%{"playlistVideoRenderer" => r}) when is_map(r) do
    with video_id when is_binary(video_id) <- r["videoId"],
         true <- Regex.match?(@video_id_re, video_id) do
      title =
        cond do
          is_list(get_in(r, ["title", "runs"])) ->
            r["title"]["runs"]
            |> Enum.map(&to_string(Map.get(&1 || %{}, "text")))
            |> Enum.join("")

          is_binary(get_in(r, ["title", "simpleText"])) ->
            r["title"]["simpleText"]

          true ->
            ""
        end

      %{"videoId" => video_id, "title" => if(is_binary(title), do: String.trim(title), else: "")}
    else
      _ -> nil
    end
  end

  defp video_from_node(%{"lockupViewModel" => l}) when is_map(l) do
    if is_binary(l["contentType"]) and l["contentType"] != "LOCKUP_CONTENT_TYPE_VIDEO" do
      nil
    else
      with video_id when is_binary(video_id) <- l["contentId"],
           true <- Regex.match?(@video_id_re, video_id) do
        title = get_in(l, ["metadata", "lockupMetadataViewModel", "title", "content"])
        %{"videoId" => video_id, "title" => if(is_binary(title), do: String.trim(title), else: "")}
      else
        _ -> nil
      end
    end
  end

  defp video_from_node(_), do: nil

  defp collect_videos(data) do
    %{videos: videos} = visit(data, %{videos: [], seen: MapSet.new()})
    videos
  end

  defp visit(node, acc) when is_list(node), do: Enum.reduce(node, acc, &visit(&1, &2))

  defp visit(node, acc) when is_map(node) do
    if length(acc.videos) >= @resolve_max do
      acc
    else
      case video_from_node(node) do
        %{"videoId" => video_id} = video ->
          if MapSet.member?(acc.seen, video_id) do
            acc
          else
            %{
              videos: acc.videos ++ [video],
              seen: MapSet.put(acc.seen, video_id)
            }
          end

        nil ->
          node
          |> Enum.sort_by(fn {k, _v} -> to_string(k) end)
          |> Enum.reduce(acc, fn {_k, v}, a -> visit(v, a) end)
      end
    end
  end

  defp visit(_node, acc), do: acc
end
